charms/knative-serving/src/charm_template.py

```diff
@@ -0,0 +1,248 @@
+    """A base charm for writing Kubernetes manifest charms
+
+    A Kubernetes manifest charm is a charm that:
+    *   derives its desired state by rendering one or more YAML templates, where these templates
+        are filled using configuration from charm configuration, relations, and hard-coded sources
+    *   deploys this desired state to Kubernetes via the Kubernetes API
+    *   monitors and, if required, reconciles these objects to the desired state
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.src_dir = Path("src")
+        self.manifests_dir = self.src_dir / "manifests"
+
+        # template_files are relative to the directory used by the jinva Environment.  By default,
+        # this is the self.manifests_dir
+        self.template_files = [
+            "manifests.yaml.j2",
+        ]
+
+        # Properties
+        self._jinja_env = None
+        self._lightkube_client = None
+
+    def charm_status(self, resources: Optional[List[Union[NamespacedResource, GlobalResource]]] = None) -> Union[ActiveStatus, WaitingStatus, BlockedStatus]:
+        """Computes status of this charm based on the state of cluster objects, logging any errors
+        """
+        self.log.info("Computing status")
+
+        try:
+            self._check_leader()
+        except LeadershipError as e:
+            return e.status
+
+        if resources is None:
+            resources = self.render_manifests()
+
+        charm_ok, errors = check_resources(self.lightkube_client, resources)
+        if charm_ok:
+            self.log.info("Status: active")
+            status = ActiveStatus()
+        else:
+            # Hit one or more errors with resources.  Return status for worst and log all
+            self.log.info("Charm is not active due to one or more issues:")
+
+            # Log all errors, ignoring None's
+            errors = [error for error in errors if error is not None]
+            for i, error in enumerate(errors):
+                self.log.info(f"Issue {i+1}/{len(errors)}: {error.msg}")
+
+            # Return status based on the worst thing we encountered
+            status = get_first_worst_error(errors).status
+
+        return status
+
+    def _check_leader(self):
+        if not self.unit.is_leader():
+            # We can't do anything useful when not the leader, so do nothing.
+            raise LeadershipError("Waiting for leadership", WaitingStatus)
+
+    def on_config_changed(self, event):
+        self.on_install(event)
+
+    def on_install(self, event):
+        """Installs all objects, checking and setting status at the end of install"""
+        self.unit.status = MaintenanceStatus("Deploying resources to cluster")
+        try:
+            self._check_leader()
+            resources = self.render_manifests()
+            self.reconcile(resources)
+        except ErrorWithStatus as e:
+            self.log_and_set_status(e.status)
+            return
```

Comment:
I think we need at the end of install to do a `self.charm_status` to determine what our status is (eg: whether we observe that the current cluster status means we're Active)